afr/heal: Implement outcast logic in metadata heal

rafikc30 opened the pull request:
When an entry heal is happening, the entry is created with
root:root permissions and none of the xattrs are set on the
entry till the meta-data heal is completed. Which means we
can say for sure that the entry is not usable till the
meta-data heal is completed. This causes problems when there
is a split-brain and favourite-child-pplicy is set. There is
a chance that the entry with wrong meta-data information
might get picked, and it's meta-data information like permissions
will be synced.

So this patch implement a way to identify the entry that
has not completed the initial meta-data heal, hence it
restricts such entries being chosen as source for the heal.

TODO:
CUrrently this logic is only implemented for the initial
meta-data heal, we could implement for data-heal as well
as ongoing modification for a better way of identifying the
source

Change-Id: Ife84bcd49e3dd99ca08f64110060edafbc9f07a5
Updates: #1546
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 committed:
afr/heal: Implement outcast logic in metadata heal

When an entry heal is happening, the entry is created with
root:root permissions and none of the xattrs are set on the
entry till the meta-data heal is completed. Which means we
can say for sure that the entry is not usable till the
meta-data heal is completed. This causes problems when there
is a split-brain and favourite-child-pplicy is set. There is
a chance that the entry with wrong meta-data information
might get picked, and it's meta-data information like permissions
will be synced.

So this patch implement a way to identify the entry that
has not completed the initial meta-data heal, hence it
restricts such entries being chosen as source for the heal.

TODO:
CUrrently this logic is only implemented for the initial
meta-data heal, we could implement for data-heal as well
as ongoing modification for a better way of identifying the
source

Change-Id: Ife84bcd49e3dd99ca08f64110060edafbc9f07a5
Updates: #1546
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

rafikc30 commented:
/run regression

rafikc30 commented:
/run regression

rafikc30 commented:
/run regression

pranithk commented on xlators/cluster/afr/src/afr-common.c:
```diff
@@ -7876,3 +7876,49 @@ afr_ta_dict_contains_pending_xattr(dict_t *dict, afr_private_t *priv, int child)
+        if (pending_raw) {
+            pending_int = pending_raw;
+
+            if (ntoh32(pending_int[idx]))
```
All we are checking is non-zero, maybe we don't need ntoh32? You can add a comment that is why you are not using ntoh32 there.

rafikc30 replied:
Done

pranithk commented on xlators/cluster/afr/src/afr-common.c:
```diff
@@ -7876,3 +7876,49 @@ afr_ta_dict_contains_pending_xattr(dict_t *dict, afr_private_t *priv, int child)
+    /*
+     * At the moment, outcast for AFR_METADATA_TRANSACTION is
+     * implemented
+     */
```
Add it as TODO before 7913 that once outcast logic is introduced for other types of heal this has to be turned into a loop.

rafikc30 replied:
Done

pranithk commented on xlators/cluster/afr/src/afr-self-heal-entry.c:
```diff
@@ -263,6 +285,10 @@ afr_selfheal_recreate_entry(call_frame_t *frame, int dst, int source,
 
     mode = st_mode_from_ia(iatt->ia_prot, iatt->ia_type);
 
+    ret = afr_set_heal_outcast(xdata, outcast, AFR_METADATA_TRANSACTION, 1);
```
Could you add a comment that, since this entry is created with this fop, no need to take metadata lock to perform this operation?

rafikc30 replied:
Done

pranithk commented on xlators/cluster/afr/src/afr-self-heal-metadata.c:
```diff
@@ -85,6 +88,9 @@ __afr_selfheal_metadata_do(call_frame_t *frame, xlator_t *this, inode_t *inode,
                 healed_sinks[i] = 0;
         }
 
+        ret = afr_set_heal_outcast(xattr, outcast, AFR_METADATA_TRANSACTION, 0);
```
I didn't see any outcast marking on bad bricks at the start of heal. Is that a deliberate choice? If yes why?

rafikc30 replied:
I'm not sure whether I understand your question correctly or not. We do mark the METADATA part of the outcast as 1 just before the recreation of an entry in the entry heal (line number 288 on the same file).

pranithk replied:
@rafikc30 It can so happen that only the metadata is modified while the brick is down. When metadata heal is triggered on that file, there won't be outcast marking. Is that intended?

rafikc30 replied:
@pranithk Since the outcast is set on the file where the healing is required, When the brick is down we won't be able to set the outcast (of course we can do this for halo replication, may be as an async fop). I believe it requires more brainstorming, so I thought may be that can be implemented along with the full outcast implementation, Where we can also consider the data part.

pranithk commented on xlators/cluster/afr/src/afr-common.c:
```diff
@@ -7876,3 +7876,49 @@ afr_ta_dict_contains_pending_xattr(dict_t *dict, afr_private_t *priv, int child)
 
     return _gf_false;
 }
+
```
Lookup should also not select iatt from outcast right? Is that change not necessary? In synchronous replication this may not be necessary because other responses would have marking. But not sure about halo.

rafikc30 replied:
Yes, I chose to do it deliberately since the lookup will read an iatt from a subvol with outcast only when all the good bricks are not available, which means if we are not reading the iatt, then basically the lookup would fail. So instead of that, I choose about giving partial data. For a user who chose halo probably aware of this risk where they can get partial data if they choose to read it from a remote dc where the min-replicas are not selected as a quorum.

So instead of providing nothing we are giving providing partial data(Same behavior as without the patch). This patch ensures that it won't then be selected as the good copy.

rafikc30 committed:
afr/outcast: Fix review comments from Pranith on 1607

Comments:
https://github.com/gluster/glusterfs/pull/1607/commits/
c0a526abf25a9422a404bb025ee5fddd5741130d#r506233938

https://github.com/gluster/glusterfs/pull/1607/commits/
c0a526abf25a9422a404bb025ee5fddd5741130d#r506236681

https://github.com/gluster/glusterfs/pull/1607/commits/
c0a526abf25a9422a404bb025ee5fddd5741130d#r506245682

Change-Id: Ib875a26ad3a814314261c745ee757f7af6defcc0
Updates: #1546
Signed-off-by: Mohammed Rafi KC <rafi.kavungal@iternity.com>

pranithk reviewed (approved):
I am okay with the changes. @itisravi @karthik-us Could you also take a look at the changes.

rafikc30 commented:
/run regression

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+TEST ls $M0/tmp1
+
+#At this point, we have completed entry-heal but not meta-data heal
+OrigUserid=$(id -u tmpuser)
```
OrigUserid seems to be generated at multiple places while it does not change. Please do it once and re-use.

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+TEST $CLI volume set $V0 cluster.metadata-self-heal off
+TEST $CLI volume set $V0 cluster.entry-self-heal on
+TEST $CLI volume set $V0 cluster.data-self-heal off
+
```
I don't see `cluster.halo-enabled` or `cluster.halo-max-replicas` etc. being set in the .t.
So is adding the outcast logic in this patch independent of halo replication? I think it should be halo specific.

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+TEST chown tmpuser:tmpuser $M0/tmp1/
+TEST chown tmpuser:tmpuser $M0/tmp1/new
+TEST chmod o-rx $M0/tmp1/
+TEST chmod o-rx $M0/tmp1/new
```
Nit: Can we chown and chmod -R  instead of on each file/dir?

itisravi commented on xlators/cluster/afr/src/afr-common.c:
```diff
@@ -7876,3 +7876,62 @@ afr_ta_dict_contains_pending_xattr(dict_t *dict, afr_private_t *priv, int child)
+    /*
+     * At the moment, outcast for AFR_METADATA_TRANSACTION is
+     * implemented
+     */
```
Both afr_is_outcast_set() and afr_is_any_outcast_set() are doing the same thing. Can you refactor so that the latter calls the former?

itisravi commented on xlators/cluster/afr/src/afr-self-heal-common.c:
```diff
@@ -161,7 +161,8 @@ afr_gfid_sbrain_source_from_src_brick(xlator_t *this, struct afr_reply *replies,
     for (i = 0; i < priv->child_count; i++) {
-        if (!replies[i].valid || replies[i].op_ret == -1)
+        if (!replies[i].valid || replies[i].op_ret == -1 ||
+            afr_is_any_outcast_set(replies[i].xdata))
```
All these changes in self-heal logic can also happen from clients (if client side heal is enabled). So if old clients are also accessing the volume, the problem can still occur no?

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+TEST pidof glusterd
+TEST $CLI volume create $V0 replica 2 $H0:$B0/${V0}{0,1}
+TEST $CLI volume set $V0 cluster.favorite-child-policy mtime
+TEST $CLI volume set $V0 cluster.self-heal-daemon off
```
Please use TEST $CLI volume heal $V0 disable/enable everywhere

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+#With this lookup from mount-point, the dht will trigger a setxattr which will force afr to do an xattrop
+#So there will be a split-brain now,
+TEST ls $M0/tmp1
+TEST $CLI volume heal $V0 disable
```
I don't think this is needed.

itisravi commented on tests/basic/afr/afr-metadata-outcast.t:
```diff
@@ -0,0 +1,70 @@
+TEST $CLI volume heal $V0 disable
+TEST $CLI volume start $V0 force
+EXPECT_WITHIN ${PROCESS_UP_TIMEOUT} "^2$" online_brick_count
+EXPECT_WITHIN ${PROCESS_UP_TIMEOUT} "^2$" afr_get_split_brain_count $V0
```
This test fails for me locally with your patch applied (I get count to be 4). I wonder why its passing here.

pranithk commented:
@rafikc30 After looking at https://github.com/gluster/glusterfs/issues/876 I realized that creating the file/directory with uid/gid of source file/dir prevents other problems too. Please also incorporate those changes into this patch.

itisravi commented:
Hey @rafikc30, are you planning to refresh this patch soon?
